File: haskell/glue/src/Glue/Lib/Math/Trigonometric/Asin.hs
module Glue.Lib.Math.Trigonometric.Asin where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Arcsine function
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Asin.asin exactly
asin :: IR Eval
asin = NativeFunc asinImpl

-- Arcsine function implementation
-- Mirrors Haskell Glue.Lib.Math.Trigonometric.Asin.asinImpl exactly
asinImpl :: IR Eval -> Eval (IR Eval)
asinImpl arg = case arg of
    Integer n -> pure $ Float (Prelude.asin (fromIntegral n))
    Float n -> pure $ Float (Prelude.asin n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/trigonometric/asin.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Arcsine function (returns radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Asin.asin exactly
final Ir asin = IrNativeFunc(asinImpl);

/// Arcsine function implementation (returns radians)
/// Mirrors Haskell Glue.Lib.Math.Trigonometric.Asin.asinImpl exactly
Eval<Ir> asinImpl(Ir arg) {
  return switch (arg) {
    IrInteger(value: final n) => Eval.pure(IrFloat(math.asin(n.toDouble()))),
    IrFloat(value: final n) => Eval.pure(IrFloat(math.asin(n))),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
